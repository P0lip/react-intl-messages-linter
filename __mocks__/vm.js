const vm = jest.genMockFromModule('vm');

module.exports = vm;
