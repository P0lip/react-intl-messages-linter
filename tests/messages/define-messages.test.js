import defineMessages from 'src/messages/define-messages';
import escodegen from 'escodegen';
import vm from 'vm';

describe('messages/define-messages', () => {
  test('ast should be exported', () => {
    expect(defineMessages.type).toBe('FunctionDeclaration');
  });

  test('ast can be consumed by escodegen', () => {
    expect(escodegen.generate(defineMessages)).toBeTruthy();
    expect(
      escodegen.generate(defineMessages).split('\n')[0],
    ).toMatchSnapshot();
  });

  test('defineMessages should store messages', () => {
    const context = {
      allMessages: new Set(),
    };
    vm.createContext(context);
    vm.runInContext(escodegen.generate(defineMessages), context);
    vm.runInContext('defineMessages({ test: { id: \'test\', defaultMessage: \'\' } })', context);
    expect([...context.allMessages]).toEqual(['test']);
    vm.runInContext('defineMessages({ foo: { id: \'bar\', defaultMessage: \'\'  } })', context);
    expect([...context.allMessages]).toEqual(['test', 'bar']);
  });

  test('defineMessages should support nested messages', () => {
    const context = {
      allMessages: new Set(),
    };

    vm.createContext(context);
    vm.runInContext(escodegen.generate(defineMessages), context);
    vm.runInContext(
      'defineMessages({ test: { Foo: { id: \'test\', defaultMessage: \'\' } }})',
      context,
    );

    expect(vm.runInContext(
      'defineMessages({ test: { Foo: { id: \'test-2\', defaultMessage: \'\' } }})',
      context,
    )).toEqual({
      test: {
        Foo: {
          id: 'test-2',
          defaultMessage: '',
        },
      },
    });
  });

  test('defineMessages should throw error if duplicated ID is encountered', () => {
    const context = {
      allMessages: new Set(),
    };
    vm.createContext(context);
    vm.runInContext(escodegen.generate(defineMessages), context);
    vm.runInContext('defineMessages({ test: { id: \'test\',  defaultMessage: \'\' } })', context);
    expect([...context.allMessages]).toEqual(['test']);
    expect(() => vm.runInContext('defineMessages({ test: { id: \'test\', defaultMessage: \'\' } })', context))
      .toThrow();
    expect([...context.allMessages]).toEqual(['test']);
  });
});
