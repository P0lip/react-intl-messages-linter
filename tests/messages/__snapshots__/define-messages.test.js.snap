// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messages/define-messages ast can be consumed by escodegen 1`] = `"function defineMessages(messages, validateDefaultMessage = true) {"`;
